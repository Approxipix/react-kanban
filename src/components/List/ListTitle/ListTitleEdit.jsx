import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { editListTitle } from '../../../redux/listReducer/actions';
import { Input, } from '../../BaseComponent';
import ClickOutside from "../../ClickOutside";

const Form = styled.form`
  width: 90%;
`;

class ListTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.listTitle || '',
    };
  }

  handleChange = ( value) => {
    this.setState({
      title: value,
    });
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.props.toggleEditTitle()
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    const { listId } = this.props;
    if (!title)  return;
    this.props.actions.editListTitle({
      listTitle: title,
      listId: listId
    });
    this.props.toggleEditTitle()
  };

  render = () => {
    const { title } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditTitle}>
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            margin="0"
            padding=".3rem .5rem"
            weight="600"
            value={title}
            placeholder="Enter list title..."
            onKeyDown={this.handleKeyDown}
            onChange={(e) => this.handleChange(e.target.value)}
            onBlur={this.handleSubmit}
            spellCheck={false}
            autoFocus
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editListTitle: editListTitle,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(ListTitleEdit);
