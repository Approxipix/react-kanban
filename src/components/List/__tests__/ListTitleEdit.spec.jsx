import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ListTitleEdit from '../ListTitle/ListTitleEdit';
import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

describe('<ListTitleEdit>', () => {
  let store;
  let component;
  let mockEditListTitle = jest.fn();
  let mockToggleEditTitle = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();

    store = mockStore({});
    store.dispatch = mockEditListTitle;
    const props = {
      listId: 'ListID',
      title: 'List Title',
      toggleEditTitle: mockToggleEditTitle
    };

    const wrapper = shallow(<ListTitleEdit store={store} {...props} />);
    component = wrapper.find('ListTitleEdit').dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should call toggleEditTitle if key code equal 27", () => {
    component.find('Input').simulate('keydown', { keyCode: 27 });

    expect(mockToggleEditTitle).toHaveBeenCalledTimes(1);
  });

  it('should call handleChange on title change with the correct params', () => {
    const spy = jest.spyOn(component.instance(), "handleChange");
    const mockEvent = {
      target: {
        name: "title",
        value: "Test title"
      }
    };

    component.find("Input").simulate("change", mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(mockEvent);
  });

  it('should call setState on title', () => {
    const mockEvent = {
      target: {
        name: "title",
        value: "Test Title"
      }
    };
    const expectedState = {
      title: "Test Title",
    };

    component.instance().handleChange(mockEvent);

    expect(component.state()).toEqual(expectedState)
  });

  it("should call handleSubmit with the correct params", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.setState({ title: "Test Title" });
    component.instance().handleSubmit(mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should return handleSubmit if title is empty", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.instance().handleSubmit(mockEvent);

    expect(spy).toReturn();
  });

  it("should dispatch an action after handleSubmit", () => {
    const expectedAction = {
      type: 'EDIT_LIST_TITLE',
      payload: {
        listTitle: 'Test Title',
        listId: 'ListID',
      }
    };

    component.setState({ title: "Test Title" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockEditListTitle).toHaveBeenCalledTimes(1);
    expect(mockEditListTitle).toHaveBeenCalledWith(expectedAction);
  });

  it("should call toggleEditTitle after submit", () => {
    component.setState({ title: "Test Title" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockToggleEditTitle).toHaveBeenCalledTimes(1);
  });
});
