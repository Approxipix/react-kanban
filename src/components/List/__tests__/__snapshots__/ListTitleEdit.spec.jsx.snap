// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListTitleEdit> should render with given state from Redux store 1`] = `
<ClickOutside
  toggleOpened={[MockFunction]}
>
  <styled.form
    onSubmit={[Function]}
  >
    <Input
      autoFocus={true}
      margin="0"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      padding=".3rem .5rem"
      placeholder="Enter list title..."
      spellCheck={false}
      type="text"
      value=""
      weight="600"
    />
  </styled.form>
</ClickOutside>
`;
